### DISCUSSION TITLE
Modifying the recorder filter at run time

### DISCUSSION THREAD

**Discussion by bdraco** [home-assistant/architecture#1106] [General] [answered]
Currently, the recorder filter can only be modified by changing YAML.  Multiple custom integrations are patching the recorder to add run-time filtering.  The patching slows Home Assistant and introduces hard-to-discover problems since the recorder runs unknown third-party code.  Because these integrations create a voluminous amount of data in the state machine, not excluding the entities will quickly fill up the database and cause different issues.

Examples of integrations patching the recorder to add excludes:
https://github.com/gcobb321/recorder_history_prefilter
https://github.com/gcobb321/icloud3
https://github.com/custom-components/places
https://github.com/enkama/hass-variables

This proposal is to add ~~a `recorder.modify_filter` service~~top level python API call to the `recorder` that would take ~~the same~~ similar arguments as https://www.home-assistant.io/integrations/recorder/#exclude and https://www.home-assistant.io/integrations/recorder/#include and another one to specify if the values should be added or removed from the filter.

The proposed implementation is at https://github.com/home-assistant/core/pull/121377

**Comment by balloob** [comment]
Should it be a service or can it be a python function?

**Reply by bdraco** [reply]
Top level python function on the recorder makes more sense for this issue.  We could always decide to make it a service later if we wanted but I think its a good idea to add only the minimal need to keep the maintenance burden as small as possible while still solving the problem

**Reply by bdraco** [reply]
Adjusted the proposal to be a python API

**Reply by gjohansson-ST** [reply]
I think it should be both but once we have the function a service is not far from. 
They serve a slightly different purpose as the function is for developers and the service would be for users?

**Reply by bdraco** [reply]
Long term, I expect this is the first step to allowing users to exclude them in the UI.  We would also want to make them persist across restarts, and have a websocket API.  However I'm getting ahead of myself here as neither of these are part of this proposal.  We could later add a service interface, however I think the websocket API would fill the need unless I've not seeing a use case.

**Comment by gcobb321** [comment]
Let me know how I can provide real world testing via iCloud3. 

@bdraco Are you using an iPhone? iCloud3 docs are [here](https://gcobb321.github.io/icloud3_v3_docs/#/chapters/0.1-introduction) If you are

**Reply by bdraco** [reply]
Once we have an accepted proposal, I will write the code to support this and than look how it can be used in icloud3

**Comment by MartinHjelmare** [comment]
How many users are there of these integrations?

**Reply by bdraco** [reply]
icloud3 seems to have quite a robust community https://community.home-assistant.io/t/icloud3-v3-idevice-tracker-version-3/267554
I'm not sure about the usage of the others.

Another motivation for this is that I've been working the recorder issue queue and spending hours investigating issues only to find they are caused by third party patching which is taking away time for solving the rest of the issues.  

ex:
https://github.com/home-assistant/core/issues/121164
https://github.com/home-assistant/core/issues/115833
https://github.com/home-assistant/core/issues/121296

Additionally every time we change how the recorder filtering works internally we end up with another round of comments about broken recorder during the release which is also happening again in 2024.7.x https://community.home-assistant.io/t/2024-7-time-to-resize-your-cards/745771 https://community.home-assistant.io/t/updating-ha-core-to-2024-7-0-energy-dasboard-is-gone-unknown-error-fixed/746022/14 https://community.home-assistant.io/t/psa-2024-7-recorder-problems/746428

**Reply by Snuffy2** [reply]
All 3 have hundreds to >1000 users (although many users overlap). For example `places` was downloaded 1700 times since 6/3/24.

FYI, for `places` and `hass-variables`, I've implemented a kinda klunky workaround by having a child class that inherits from my main entity class that sets `_unrecorded_attributes = frozenset({MATCH_ALL})` as a class attribute. Then in `async_setup_entry` calling `async_add_entities` for the normal vs the no-recorder class.

https://github.com/custom-components/places/pull/290 - **Live in v2.7.1**
https://github.com/enkama/hass-variables/pull/121 - **Live in v3.4.4**

**Reply by gcobb321** [reply]
ICloud3 v3 has just recently been released with a slow rollout and probably has ~100.
iCloud3 v2 has a GitHub download count of 13,000+

**Comment by gcobb321** [comment]
One approach for custom components might be to create a new sensor attribute in the sensor’s SensorEntity class (__init__.py) such as _attr_recorder_exclude (True/False) that is checked in the recorder component core.py to see if the sensor is excluded or not. A custom component could set an _attr_recorder_exclude property in its sensor.py  to True if it should be excluded. 

This might be simpler that having the custom component update the exclude list in the Recorder module. Just a thought.

**Comment by bdraco** [comment]
The proposed implementation is at https://github.com/home-assistant/core/pull/121377

**Comment by frenck** [comment]
Looking at the central role/importance of recording on Home Assistant, I think we could say it is an integral part of the core of Home Assistant. From that perspective, I think we should move these properties out of the recorder, but attach this information to the objects stored.

Which means: Let's the entities hold information about if they should be recorded or not. With that in mind, a UI for it should become pretty simple as well (as we have entity settings).

**Reply by bdraco** [reply]
Are you thinking the recorder/logbook/history filtering functionality should be removed and replaced with entity settings? If so, how would we handle domain and glob include/excludes?

**Reply by PeteRager** [reply]
I like that a lot.  Being able to add recorder config to individual entities would be great examples include:

- deadband - how much of a change is required before recording. Example, recording voltages, I don't care if they are oscillating by 0.1 v; but I do want to see 0.1v in states.  So I'd set the deadband to 1v.
- time - how frequently to log. Example, home power readings.   I read these every 10 seconds from modbus.  I then have a separate trigger based sensor on a 1 minute trigger which I record.  Instead I'd set the time to 1 minute
- retention - how long to retain.   As an example zwave diagnostic data, I purge after 10 days, on the other hand data from my water wells I want to keep the raw data indefinitely - as I'm still trying to build a ground water forecaster.  Since MariaDB scales fine on my hardware; I'm not using influx for long term storage as the recorder graphing doesn't automatically switch between long and short term storage. 

For backwards compatibility will should still support the existing config.  We'd need to decide if that gets applied before or after the entity specific config.  I think the entity config should win.

another approach I've seen is by defining recording groups.   Each group has specific config. As an example, I could define a group called "voltages", include any sensors of device class voltage, and specify a deadband of 1 volt. The advantage is it scales better - as I add new voltage sensors - they automatically get the right config.  This is more aligned with the current config.  If we are this, I'd expect to not have more than 10-15 groups defined.  The current config could be treated as the "default" group.

**Reply by gjohansson-ST** [reply]
I would like this approach that perhaps we maintain the current functionality which supports domain and glob and also entities which are not stored (no unique id) but we add an entity option which is then stored by the entity itself.
Then we could get support for `suggest_recording` from integrations (developers) and in the end the user can themselves set it per entity.

**Reply by kroimon** [reply]
I like the idea of recordability being a property of an entity and of users being able to more easily configure this on a per-entity basis - matching two of HA's core principles: **privacy** (don't record sensitive data) and **choice** (let the user decide what to record, not some integration).
Having even more properties like deadband, time interval or retention time would be great future enhancements, too, and I think those could only ever be implemented as entity properties.

> Are you thinking the recorder/logbook/history filtering functionality should be removed and replaced with entity settings? If so, how would we handle domain and glob include/excludes?

@bdraco Could you elaborate on how this would affect domain and glob includes/excludes?

**Reply by bdraco** [reply]
If we attached it to entities/objects themselves we wouldn't be able to support domains/globs anymore in the backend.  Instead the UI would have to offer methods to batch select and configure.  Adding a UI component would vastly increase the scope here so we would need to do some collaboration with the frontend team to make that happen as it would affect a lot more places.

**Reply by PeteRager** [reply]
We should not break all existing recorder configuration.  Whatever we do should be additive.  Perhaps adding one more level such that you can define a "recording group" that has globs (e.g. what we have now) underneath it.  In the group you could specify retention time, deadbands, etc.   We treat the existing config as belonging to the default group.

**Reply by gjohansson-ST** [reply]
> If we attached it to entities/objects themselves we wouldn't be able to support domains/globs anymore in the backend.

Why couldn't we do both?
Control specific entities from entity options and then more advanced configuration still remain as today including configuring domains/globs.

**Reply by bdraco** [reply]
We would have two sources of truth. In this case which filter would win?

**Reply by gjohansson-ST** [reply]
I think it should work just like today. By default entities are included and if set as excluded the get "appended" to the excluded entities list. 
Example from the recorder page
```yaml
# Example configuration.yaml entry with include and exclude
recorder:
  include:
    domains:
      - sensor
      - switch
      - media_player
  exclude:
    entities:
      - sensor.last_boot
      - sensor.date
      - ... [add more from entity options]
    entity_globs:
      - sensor.weather_*
```

Obviously only do the calculations and not change the yaml configuration.

**Comment by balloob** [comment, marked as answer]
This feature has been approved in https://github.com/home-assistant/architecture/discussions/1276#discussioncomment-14307564 . It is limited to doing a one-time migration from YAML to a recorder registry that can be adjusted at runtime.
